Fix SquashFS extraction when it contains directory symlinks

Flamefire opened the pull request:
## Description of the Pull Request (PR):

A SquashFS image might contain directory symlinks such as `/var/tmp -> /tmp`.
Before extraction a base environment is created (`makeBaseEnv`) by pre-creating some directories. Among them `/var/tmp`.
Although unsquashfs is called with `-force` it fails to replace that by the symlink in the image as it removes only directories.

Fix by creating the base env after the extracion as done for e.g. OCI images.


### This fixes or addresses the following GitHub issues:

 - Fixes #2347


Based on https://github.com/sylabs/singularity/pull/3151

Flamefire commented:
Neither of the failures seem to be caused by this PR:

- `FAIL: TestE2E/PAR/KEY/ordered/keyCmd/key_search_-u_https://keys.openpgp.org_0x<key_fingerprint>`
- RPM build fails to install system package

DrDaveD reviewed (changes requested):
Thanks for the submission!

I noticed that you are listed as the author of all the commits, even though some of the commits in the SingularityCE PR were committed by someone else.  It's important to the SingularityCE folks that authorship is maintained.  So please-redo the commits either as cherry-picks which automatically retain authorship, or do the commits with the `--author` option listing the original author.

DrDaveD commented:
> Neither of the failures seem to be caused by this PR:
> 
>     * `FAIL: TestE2E/PAR/KEY/ordered/keyCmd/key_search_-u_https://keys.openpgp.org_0x<key_fingerprint>`
> 
>     * RPM build fails to install system package

I re-ran the failed tests and they succeeded

Flamefire commented:
> I noticed that you are listed as the author of all the commits, even though some of the commits in the SingularityCE PR were committed by someone else. 

This is a bit tricky.

- The actual fix is fully mine (they only fixed a merge conflict as I initially tried it here).
- The test case is a mix: The initial version is mine, there were some review comments and later a refactoring from them. I used my version for this PR but incorporated those review comments and parts of the refactoring.
- The E2E test is mostly from there but I changed some parts to adjust for apptainer. I could change the author of that commit and just mention that I adjusted it.

For the latter 2:  Shall I make 2 commits? The singularity commit and one basically resetting it to my version to adjust for apptainer which isn't fully trivial as e.g. the makesquashfs function isn't available here

DrDaveD commented:
Sure you can make the latter 2 separate commits.  Or it's fine to edit their commit a bit and leave them as an author.  Github shows both the one marked author and the committer so it's clear that it was a 2-person effort.

Flamefire commented:
> Github shows both the one marked author and the committer so it's clear that it was a 2-person effort.

More or less: A cherry-pick without any modification does look the same.

What I did now:
- Add my test commit
- add part of their commit related to that commit and a followup
- Reset to the version I added here in the next commit
- Add my fix commit
- add the other part of their commit adding the E2E test
- Reset to my version of that test in the next commit

I think now it should be clear to whom each change belongs.

DrDaveD commented:
Thanks!  Please cherry-pick the commits in this PR to another PR against the release-1.3 branch.

Flamefire commented:
Done: https://github.com/apptainer/apptainer/pull/2400
